// comicvine/resource.js
var Resource = {};
var API_KEY = '42f916cdc6ab0316b376f4a46731bcabf7463fb9';
var XHR = require('./xhr');

Resource.RESOURCES_CODES = {
  characters: 4005,
  teams: 4060,
  publishers: 4010,
  concepts: 4015,
  issues: 4000,
  movies: 4025,
  episodes: 4070,
  story_arcs: 4045,
  creators: 4040,
  things: 4055,
  locations: 4020,
  volumes: 4050,
  series: 4075
}

Resource.find = function(options) {
  var field_list = (options.field_list || []).join(',');

  return XHR.get('http://www.comicvine.com/api/' + options.resourceName + '/' + options.resourceCode + '-' + options.id + '/', {
    api_key: API_KEY,
    format: 'json',
    field_list: field_list
  }).then(function(data){
    return data.results;
  });
};

Resource.filter = function(options) {
  var field_list = (options.field_list || []).join(',');
  
  var filter = Object.keys(options.where).map(function(key) {
    return key + ':' + options.where[key];
  }).join(',');

  return XHR.get('http://www.comicvine.com/api/' + options.resourceName + '/', {
    api_key: API_KEY,
    filter: filter,
    format: 'json',
    field_list: field_list
  }).then(function(data){
    return data.results;
  });
};

module.exports = Resource;
