// coffre/associations.js
var Inflecta = require('inflecta'),
    Coffre = require('./index'),
    Relation = require('./relation');
var Associations = {};

Associations.hasMany = function(tableName, options) {
  options = options || {};

  var propertyModel = Coffre.mappedModels[Inflecta.tableize(tableName)] || options['model'],
      propertyName = options['as'] || tableName;

  this.associations.hasMany[tableName] = {
    model: propertyModel,
    as: propertyName
  };

  Object.defineProperty(this.prototype, propertyName, {
    configurable: false,
    enumerable: true,
    get: function() {
      if (this.associations.hasMany[tableName] === undefined) {
        if (this.constructor.associations.hasMany[tableName].model === undefined) {
          this.constructor.associations.hasMany[tableName].model = Coffre.mappedModels[Inflecta.tableize(tableName)];
        }

        var conditions = {};
        conditions[Inflecta.singularize(this.constructor.tableName) + '_id'] = this.id;

        this.associations.hasMany[tableName] = new Relation({
          model: Coffre.mappedModels[Inflecta.tableize(tableName)],
          table: tableName,
          values: {
            where: conditions
          }
        });
      }

      return this.associations.hasMany[tableName];
    }
  });
};

Associations.belongsTo = function(tableName, options) {
  options = options || {};

  var propertyModel = Coffre.mappedModels[Inflecta.tableize(tableName)] || options['model'],
      propertyName = options['as'] || tableName;

  this.associations.belongsTo[tableName] = {
    model: propertyModel,
    as: propertyName
  };

  Object.defineProperty(this.prototype, propertyName, {
    configurable: false,
    enumerable: true,
    get: function() {
      if (this.associations.belongsTo[tableName] === undefined) {
        if (this.constructor.associations.belongsTo[tableName].model === undefined) {
          this.constructor.associations.belongsTo[tableName].model = Coffre.mappedModels[Inflecta.tableize(tableName)];
        }

        var conditions = {};
        conditions['id'] = this[Inflecta.singularize(tableName) + '_id'];

        this.associations.belongsTo[tableName] = new Relation({
          model: Coffre.mappedModels[Inflecta.tableize(tableName)],
          table: tableName,
          values: {
            where: conditions,
            oneResult: true
          }
        });
      }

      return this.associations.belongsTo[tableName];
    }
  });
};

module.exports = Associations;
