// coffre/model.js
var Inflecta = require('inflecta'),
    Coffre = require('./index'),
    Querying = require('./querying'),
    Persistence = require('./persistence'),
    Associations = require('./associations');

var Model = {};

Model.defineSchema = function(model) {
  var schema = JSON.parse(global.window.localStorage.getItem('schema:' + model.tableName)) || [];
  model.attributes = [];

  schema.forEach(function(column) {
    model.attributes.push(column['name']);
    
    Object.defineProperty(model.prototype, column['name'], {
      configurable: false,
      enumerable: true,
      get: function() {
        return this.attributes[column['name']] || null;
      },
      set: function(value) {
        this.attributes[column['name']] = value;
      }
    });
  });
};

Model.define = function(modelName, modelOptions) {
  var newModel = function Model(attributes) {
    this.attributes = {};

    for (var i in attributes) {
      this.attributes[i] = attributes[i];
    }

    if (typeof this.attributes.createdAt === 'number') {
      this.attributes.createdAt = new Date(this.attributes.createdAt);
    }

    if (typeof this.attributes.updatedAt === 'number') {
      this.attributes.updatedAt = new Date(this.attributes.updatedAt);
    }
  };

  newModel.tableName = Inflecta.tableize(modelName);
  Coffre.mappedModels[newModel.tableName] = newModel;

  Model.defineSchema(newModel);

  Object.keys(Querying.staticMethods).forEach(function(staticMethod) {
    newModel[staticMethod] = Querying.staticMethods[staticMethod];
  });

  Object.keys(Persistence.staticMethods).forEach(function(staticMethod) {
    newModel[staticMethod] = Persistence.staticMethods[staticMethod];
  });

  Object.keys(Persistence.instanceMethods).forEach(function(instanceMethod) {
    newModel.prototype[instanceMethod] = Persistence.instanceMethods[instanceMethod];
  });

  newModel.prototype.toJSON = function() {
    var json = {};

    var attributesKeys = Object.keys(this.attributes),
        hasManyKeys = Object.keys(this.associations.hasMany),
        belongsToKeys = Object.keys(this.associations.belongsTo);

    for (var i = 0; i < attributesKeys.length; i++) {
      json[attributesKeys[i]] = this.attributes[attributesKeys[i]];
    }

    for (var i = 0; i < hasManyKeys.length; i++) {
      json[hasManyKeys[i]] = this.associations.hasMany[hasManyKeys[i]].map(function(item) {
        return item.toJSON();
      });
    }

    for (var i = 0; i < belongsToKeys.length; i++) {
      json[belongsToKeys[i]] = this.associations.belongsTo[belongsToKeys[i]][0].toJSON();
    }

    return json;
  };

  if (modelOptions instanceof Function) {
    newModel.associations = {
      hasMany: {},
      belongsTo: {}
    };

    newModel.prototype.associations = {
      hasMany: {},
      belongsTo: {}
    };
    
    newModel.hasMany = Associations.hasMany;
    newModel.belongsTo = Associations.belongsTo;

    modelOptions.call(newModel);
  }

  return newModel;
};

module.exports = Model;
